sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/Metrics.java

```diff
@@ -0,0 +1,234 @@
+package org.apache.beam.fn.harness.control;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.beam.fn.harness.control.ProcessBundleHandler.BundleProcessor;
+import org.apache.beam.runners.core.metrics.DistributionData;
+import org.apache.beam.runners.core.metrics.MonitoringInfoEncodings;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Distribution;
+import org.apache.beam.sdk.metrics.MetricName;
+import org.apache.beam.vendor.grpc.v1p43p2.com.google.protobuf.ByteString;
+
+public class Metrics {
+
+  /**
+   * A {@link Counter} that is designed to report intermediate and final results and can be re-used
+   * after being reset.
+   */
+  public interface BundleCounter extends BundleProgressReporter, Counter {}
+
+  /**
+   * A {@link Distribution} that is designed to report intermediate and final results and can be
+   * re-used after being reset.
+   */
+  public interface BundleDistribution extends BundleProgressReporter, Distribution {}
+
+  /**
+   * Returns a counter that will report an accurate final value.
+   *
+   * <p>All invocations to {@link Counter} methods, {@link
+   * BundleProgressReporter#updateFinalMonitoringData}, and {@link BundleProgressReporter#reset()}
+   * must be done by the same thread.
+   *
+   * <p>All invocations to {@link BundleProgressReporter} methods must be done serially.
+   *
+   * <p>Invocations to {@link Counter} methods can be done concurrently to {@link
+   * BundleProgressReporter#updateIntermediateMonitoringData}.
+   */
+  public static BundleCounter serialMutatorAndFinalReaderBundleCounter(
+      String shortId, MetricName name) {
+    return new SerialMutatorAndFinalReaderBundleCounter(shortId, name);
+  }
+
+  /**
+   * Returns a {@link Distribution} that will report an accurate final value.
+   *
+   * <p>All invocations to {@link Distribution} methods, {@link
+   * BundleProgressReporter#updateFinalMonitoringData}, and {@link BundleProgressReporter#reset()}
+   * must be done by the same thread.
+   *
+   * <p>All invocations to {@link BundleProgressReporter} methods must be done serially.
+   *
+   * <p>Invocations to {@link Distribution} methods can be done concurrently to {@link
+   * BundleProgressReporter#updateIntermediateMonitoringData}.
+   */
+  public static BundleDistribution serialMutatorAndFinalReaderBundleDistribution(
+      String shortId, MetricName name) {
+    return new SerialMutatorAndFinalReaderBundleDistribution(shortId, name);
+  }
+
+  /**
+   * A {@link Counter} that shares the data using {@link AtomicReference#lazySet} to reduce
+   * synchronization overhead. This guarantees that intermediate progress reports will report a
+   * value that has been recently updated while the final progress report will get the true final
+   * value.
+   */
+  @NotThreadSafe
```

Comment:
It is still not thread safe since a lock must be hold when invoking this as the lastReportedValue is unguarded.

Reply:
The thread model of this is a little confusing. It appears the BundleProgressReporter and Counter interface can be called from different threads but otherwise aren't thread safe. Merging the two interfaces muddles that a bit and I fear it will eventually get misused.

Reply:
I think the issue is that it isn't thread safe to be called from arbitrary threads but is thread safe if used under a specific pattern.

The thread interaction boils down to who owns:
* (A) updating the counter
* (B) reading intermediate progress
* (C) reading final progress
* (D) performing the reset

The currently added counter requires A, C, D to be done via one thread or under a lock and B, C, D to be done either via one thread or while under a lock. The current setup has the main processing thread do A, C, D since it is one thread that does these always and B, C, D via lock.

To address your concerns I simplified the comments down to how to use vs the general thread safety semantics which should help a lot in having them not used inappropriately.